txnsync/mainloop.go

```diff
@@ -0,0 +1,436 @@
+			profOutgoingMsg.end()
+			continue
+		case <-s.nextOffsetRollingCh:
+			profNextOffset.start()
+			s.rollOffsets()
+			profNextOffset.end()
+			continue
+		case <-serviceCtx.Done():
+			return
+		default:
+		}
+
+		profIdle.start()
+		select {
+		case ent := <-externalEvents:
+			profIdle.end()
+			switch ent.eventType {
+			case transactionPoolChangedEvent:
+				profTxChange.start()
+				s.onTransactionPoolChangedEvent(ent)
+				profTxChange.end()
+			case newRoundEvent:
+				profNewRounnd.start()
+				s.onNewRoundEvent(ent)
+				profNewRounnd.end()
+			}
+		case <-nextPeerStateCh:
+			profIdle.end()
+			profPeerState.start()
+			s.evaluatePeerStateChanges(nextPeerStateTime)
+			profPeerState.end()
+		case incomingMsg := <-s.incomingMessagesQ.getIncomingMessageChannel():
+			profIdle.end()
+			profIncomingMsg.start()
+			s.evaluateIncomingMessage(incomingMsg)
+			profIncomingMsg.end()
+		case msgSent := <-s.outgoingMessagesCallbackCh:
+			profIdle.end()
+			profOutgoingMsg.start()
+			s.evaluateOutgoingMessage(msgSent)
+			profOutgoingMsg.end()
+		case <-s.nextOffsetRollingCh:
+			profIdle.end()
+			profNextOffset.start()
+			s.rollOffsets()
+			profNextOffset.end()
+		case <-serviceCtx.Done():
+			profIdle.end()
+			return
+		case <-s.node.NotifyMonitor():
+			profIdle.end()
+		}
+	}
+}
+
+func (s *syncState) onTransactionPoolChangedEvent(ent Event) {
+	if ent.transactionHandlerBacklogFull {
+		// if the transaction handler backlog is full, we don't want to receive any more transactions.
+		// setting the transactionPoolFull here would notify other nodes that we don't want any more messages.
+		s.transactionPoolFull = true
+	} else if s.transactionPoolFull {
+		// the transaction pool is currently full.
+		if float32(ent.transactionPoolSize) < float32(s.config.TxPoolSize)*transactionPoolLowWatermark {
+			s.transactionPoolFull = false
+		}
+	} else {
+		if float32(ent.transactionPoolSize) > float32(s.config.TxPoolSize)*transactionPoolHighWatermark {
+			s.transactionPoolFull = true
+		}
+	}
+
+	newBeta := beta(ent.transactionPoolSize)
+
+	// check if beta should be updated
+	if !shouldUpdateBeta(s.lastBeta, newBeta, betaGranularChangeThreshold) {
+		// no changes
+		return
+	}
+	// yes, change beta as the number of transactions in the pool have changed dramatically since the last time.
+	s.lastBeta = newBeta
+
+	peers := make([]*Peer, 0, len(s.interruptablePeers))
+	for _, peer := range s.interruptablePeers {
+		if peer == nil {
+			continue
+		}
+		peers = append(peers, peer)
+		peer.state = peerStateHoldsoff
+	}
+
+	// reset the interruptablePeers array, since all it's members were made into holdsoff
+	s.interruptablePeers = nil
+	s.interruptablePeersMap = make(map[*Peer]int)
+	deadlineMonitor := s.clock.DeadlineMonitorAt(s.clock.Since() + sendMessagesTime)
+	s.sendMessageLoop(s.clock.Since(), deadlineMonitor, peers)
+
+	currentTimeout := s.clock.Since()
+	for _, peer := range peers {
+		peerNext := s.scheduler.peerDuration(peer)
```

Comment:
I think it's a good point - could you open a ticket for that ?
I don't know that it's currently an issue - but it would be a good idea to address that.

Reply:
https://github.com/algorand/go-algorand-internal/issues/1507